pkg/blueprint/partitioning_customizations.go

```diff
@@ -0,0 +1,261 @@
+package blueprint
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"slices"
+	"strings"
+
+	"github.com/osbuild/images/pkg/pathpolicy"
+)
+
+type PartitioningCustomization struct {
+	MinSize uint64                        `json:"minsize,omitempty" toml:"minsize,omitempty"`
+	Plain   *PlainFilesystemCustomization `json:"plain,omitempty" toml:"plain,omitempty"`
+	LVM     *LVMCustomization             `json:"lvm,omitempty" toml:"lvm,omitempty"`
+	Btrfs   *BtrfsCustomization           `json:"btrfs,omitempty" toml:"btrfs,omitempty"`
+}
+
+type PlainFilesystemCustomization struct {
+	Filesystems []FilesystemCustomization `json:"filesystems,omitempty" toml:"filesystems,omitempty"`
+}
+
+type LVMCustomization struct {
+	VolumeGroups []VGCustomization `json:"volume_groups,omitempty" toml:"volume_groups,omitempty"`
+}
+
+type VGCustomization struct {
+	// Volume group name
+	Name string `json:"name" toml:"name"`
+	// Size of the partition that contains the volume group
+	MinSize        uint64            `json:"minsize" toml:"minsize"`
+	LogicalVolumes []LVCustomization `json:"logical_volumes,omitempty" toml:"logical_volumes,omitempty"`
+}
+
+type LVCustomization struct {
+	// Logical volume name
+	Name string `json:"name,omitempty" toml:"name,omitempty"`
+	FilesystemCustomization
+}
+
+type BtrfsCustomization struct {
+	Volumes []BtrfsVolumeCustomization
+}
+
+type BtrfsVolumeCustomization struct {
+	// Size of the btrfs partition/volume.
+	MinSize    uint64 `json:"minsize" toml:"minsize"`
+	Subvolumes []BtrfsSubvolumeCustomization
+}
+
+type BtrfsSubvolumeCustomization struct {
+	Name       string `json:"name" toml:"name"`
+	Mountpoint string `json:"mountpoint" toml:"mountpoint"`
+}
+
+func validateMountpoint(path string) error {
+	if path == "" {
+		return fmt.Errorf("mountpoint is empty")
+	}
+
+	if !strings.HasPrefix(path, "/") {
+		return fmt.Errorf("mountpoint %q is not an absolute path", path)
+	}
+
+	if cleanPath := filepath.Clean(path); path != cleanPath {
+		return fmt.Errorf("mountpoint %q is not a canonical path (did you mean %q?)", path, cleanPath)
+	}
+
+	return nil
+}
+
+func validateFilesystemType(path, fstype string) error {
+	// Check that the fs type is valid for the mountpoint.
+	// Empty strings are allowed for fstype to set the type automatically based
+	// on the distro defaults.
+	badfsMsg := "unsupported filesystem type for %q: %s"
+	switch path {
+	case "/boot":
+		switch fstype {
+		case "xfs", "ext4", "":
+		default:
+			return fmt.Errorf(badfsMsg, path, fstype)
+		}
+	case "/boot/efi":
+		switch fstype {
+		case "vfat", "":
+		default:
+			return fmt.Errorf(badfsMsg, path, fstype)
+		}
+	}
+	return nil
+}
+
+// ValidateSimple checks that at most one LVM Volume Group or btrfs volume is
+// defined. Returns an error if both LVM and btrfs are set and if either has
+// more than one element.
+// This function also calls [PartitioningCustomization.Validate].
+func (p *PartitioningCustomization) ValidateSimple() error {
```

Comment:
This is 100% a naming issue.  Validate simple sounds like it does a simpler form of validation, but instead it's meant to imply that it's requiring a simpler set of customizations (no mixing btrfs with lvm for example).  We need a better name for it.  Something that says "Validate but apply the stricter rules as well" (1 volume, no mixing).

Reply:
Right, that makes sense from an implementation point of view. I wonder if from an API/consumer point of view it still wouldn't make sense to flip it around, i.e. as a user of this API I want to call `PartitioningCustomization.Validate()` - as an (API) user the fact that we (somewhat artificially) limit what can be used does not matter, all I want is a (single) validate (I think).

So maybe something like: 
```go
diff --git a/pkg/blueprint/partitioning_customizations.go b/pkg/blueprint/partitioning_customizations.go
index 280f745bf..da1b260df 100644
--- a/pkg/blueprint/partitioning_customizations.go
+++ b/pkg/blueprint/partitioning_customizations.go
@@ -92,11 +92,10 @@ func validateFilesystemType(path, fstype string) error {
        return nil
 }
 
-// ValidateSimple checks that at most one LVM Volume Group or btrfs volume is
-// defined. Returns an error if both LVM and btrfs are set and if either has
-// more than one element.
-// This function also calls [PartitioningCustomization.Validate].
-func (p *PartitioningCustomization) ValidateSimple() error {
+// validateLayoutConstraints checks that at most one LVM Volume Group
+// or btrfs volume is defined. Returns an error if both LVM and btrfs
+// are set and if either has more than one element.
+func (p *PartitioningCustomization) validateLayoutConstraints() error {
        if p == nil {
                return nil
        }
@@ -113,7 +112,7 @@ func (p *PartitioningCustomization) ValidateSimple() error {
                return fmt.Errorf("multiple LVM volume groups are not yet supported")
        }
 
-       return p.Validate()
+       return nil
 }
 
 // Validate checks for customization combinations that are generally not
@@ -124,6 +123,10 @@ func (p *PartitioningCustomization) Validate() error {
                return nil
        }
 
+       if err := p.validateLayoutConstraints(); err != nil {
+               return err
+       }
+
        // iterate through everything and look for:
        // - invalid mountpoints (global)
        // - duplicate mountpoints (global)
diff --git a/pkg/blueprint/partitioning_customizations_test.go b/pkg/blueprint/partitioning_customizations_test.go
index d469acc6a..de61be42a 100644
--- a/pkg/blueprint/partitioning_customizations_test.go
+++ b/pkg/blueprint/partitioning_customizations_test.go
@@ -660,7 +660,7 @@ func TestPartitioningValidation(t *testing.T) {
        for name := range testCases {
                tc := testCases[name]
                t.Run(name, func(t *testing.T) {
-                       err := tc.partitioning.ValidateSimple()
+                       err := tc.partitioning.Validate()
                        if tc.expectedMsg != "" {
                                assert.EqualError(t, err, tc.expectedMsg)
                        } else {
```
(names TBD) - but maybe I'm just missing something here?

Reply:
The problem I was trying to solve is being able to validate without the layout constraints, so that the blueprint code itself applies the extra constraints (at the UI level) but the internals (`NewCustomPartitionTable()`) only check for invalid configurations like conflicting names and labels.

I think I'm going to take your suggestion but keep them separate, make `validateLayoutConstraints()` private like you suggest and run it completely separately.  So neither function would call the other.